### PR TITLE
Fix#1312 - validate query params for GET /query?text=<searchTerm>&filter=<filterBy> route

### LINKED ISSUE #1312: Add query params validation for GET /query?text=<searchTerm>&filter=<filterBy> route using Express Validator and unit tests
**What would you like to be added**: 
Request to add query validation for GET /query?text=<searchTerm>&filter=<filterBy> route using [Express-validator](https://express-validator.github.io/docs/) and also unit tests to ensure that it works

**Why would you like this to be added**:
It would be better to validate the query before it reach the route handler. 

reference: https://express-validator.github.io/docs/check-api.html

All logic for the middleware should be place inside `backend/web/validation.js`

### PR DESCRIPTION
**Pull request by chawlapalak**
Fixes #1312 

## Type of Change

- [ ] **Bugfix**: Change which fixes an issue
- [x] **New Feature**: Change which adds functionality
- [ ] **Documentation Update**: Change which improves documentation
- [ ] **UI**: Change which improves UI

## Description
 

- Added query validation for text and filter in GET/query using Express-validator.
-  Also added unit tests to ensure that it works.

## Checklist

- [x] **Quality**: This PR builds and passes our npm test and works locally
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not (if applicable)
- [x] **Documentation**: This PR includes updated/added documentation to user exposed functionality or configuration variables are added/changed or an explanation of why it does not(if applicable)

### CODE REVIEW

File: src/backend/web/routes/query.js [hunk 1]
```diff
@@ -1,29 +1,14 @@
 const express = require('express');
 
+const { queryValidationRules, validateQuery } = require('../validation');
 const { logger } = require('../../utils/logger');
 const { search } = require('../../utils/indexer');
 
 const router = express.Router();
 
-router.get('/', async (req, res) => {
+router.get('/', queryValidationRules(), validateQuery, async (req, res) => {
```
Comment (humphd) [inline src/backend/web/routes/query.js:9]: Let's not leak implementation details of the validation here by having two separate middleware functions.  In the `validation.js` file, you should amalgamate these two into a single `validateQuery()` function.  Also, the convention we have been using with middleware functions is to call them here and have them return a middleware function (vs. being a function already).

File: src/backend/web/validation.js [hunk 2]
```diff
@@ -50,4 +51,41 @@ function validateNewFeed() {
   };
 }
 
-module.exports.validateNewFeed = validateNewFeed;
+// query validation starts here
+// queryValidation rules
+const queryValidationRules = () => {
```
Comment (humphd) [inline src/backend/web/validation.js:56]: Fold this into the `validateQuery` function below, so we don't expose it outside this module.

File: src/backend/web/validation.js [hunk 3]
```diff
@@ -50,4 +51,41 @@ function validateNewFeed() {
...
+  return [
+    // text must be between 1 and 256
+    check('text')
+      .exists()
+      .withMessage('text should not be empty')
+      .isLength({ max: 256, min: 1 })
+      .withMessage('text should be between 0 to 256 characters'),
```
Comment (humphd) [inline src/backend/web/validation.js:63]: You use 1 to 256 above, but specify 0 to 256 here.  Let's be consistent.

File: src/backend/web/validation.js [hunk 4]
```diff
@@ -50,4 +51,41 @@ function validateNewFeed() {
...
+    // filter must exists
```
Comment (humphd) [inline src/backend/web/validation.js:64]: exist vs. exists

File: src/backend/web/validation.js [hunk 5]
```diff
@@ -50,4 +51,41 @@ function validateNewFeed() {
...
+};
+
+// queryValidationFunction
+const validateQuery = (req, res, next) => {
+  const errors = validationResult(req);
+  if (errors.isEmpty()) {
+    return next();
+  }
+  const extractedErrors = [];
+  errors.array().map((err) => extractedErrors.push({ [err.param]: err.msg }));
```
Comment (humphd) [inline src/backend/web/validation.js:80]: `map()` returns an array, so you don't need to `push()` manually:

```js
const extracted Errors = errors.array().map((err) => ({ [err.param]: err.msg }));
```

File: test/query.test.js [hunk 6]
```diff
@@ -3,14 +3,14 @@ const app = require('../src/backend/web/app');
 
 jest.mock('../src/backend/utils/indexer');
 
-describe('Testing query route', () => {
-  test('Testing with valid length of characters', async () => {
+describe('Testing query route with', () => {
```
Comment (humphd) [inline test/query.test.js:6]: `with` what?  I'm not clear on why you added this.

File: test/query.test.js [hunk 7]
```diff
@@ -19,4 +19,27 @@ describe('Testing query route', () => {
...
+    expect(res.status).toBe(400);
+  });
+
+  test('Testing with invalid filter', async () => {
+    const searchTerm = encodeURIComponent('I Love Telescope');
+    const res = await request(app).get(`/query?text=${searchTerm}&filter=test`);
+    expect(res.status).toBe(400);
+  });
+
+  test('Testing with valid filter', async () => {
```
Comment (humphd) [inline test/query.test.js:40]: Let's do another test like this, but for `author` so we have both.

**Review by humphd — CHANGES_REQUESTED** [review]

File: test/query.test.js (hunk 7 above)
Comment (chawlapalak) [reply test/query.test.js]: Incorporated in [7dd761b](https://github.com/Seneca-CDOT/telescope/pull/1334/commits/7dd761bc7b9281cd31a572a30424e0c41235cb6b)

File: test/query.test.js (hunk 6 above)
Comment (chawlapalak) [reply test/query.test.js]: Incorporated in [7dd761b](https://github.com/Seneca-CDOT/telescope/pull/1334/commits/7dd761bc7b9281cd31a572a30424e0c41235cb6b)

File: src/backend/web/validation.js (hunk 5 above)
Comment (chawlapalak) [reply src/backend/web/validation.js]: Incorporated in [7dd761b](https://github.com/Seneca-CDOT/telescope/pull/1334/commits/7dd761bc7b9281cd31a572a30424e0c41235cb6b)

File: src/backend/web/validation.js (hunk 4 above)
Comment (chawlapalak) [reply src/backend/web/validation.js]: Incorporated in [7dd761b](https://github.com/Seneca-CDOT/telescope/pull/1334/commits/7dd761bc7b9281cd31a572a30424e0c41235cb6b)

File: src/backend/web/validation.js (hunk 3 above)
Comment (chawlapalak) [reply src/backend/web/validation.js]: Incorporated in [7dd761b](https://github.com/Seneca-CDOT/telescope/pull/1334/commits/7dd761bc7b9281cd31a572a30424e0c41235cb6b)

File: src/backend/web/validation.js (hunk 2 above)
Comment (chawlapalak) [reply src/backend/web/validation.js:56]: Incorporated in [7dd761b](https://github.com/Seneca-CDOT/telescope/pull/1334/commits/7dd761bc7b9281cd31a572a30424e0c41235cb6b)

File: src/backend/web/routes/query.js (hunk 1 above)
Comment (chawlapalak) [reply src/backend/web/routes/query.js]: Incorporated in [7dd761b](https://github.com/Seneca-CDOT/telescope/pull/1334/commits/7dd761bc7b9281cd31a572a30424e0c41235cb6b)

**Review by zg3d — APPROVED** [review]
Looks good to me.

File: test/query.test.js [hunk 8]
```diff
@@ -19,4 +19,38 @@ describe('Testing query route', () => {
...
+  test('Testing with valid author filter', async () => {
+    const searchTerm = encodeURIComponent('I Love Telescope');
+    const res = await request(app).get(`/query?text=${searchTerm}&filter=author`);
+    expect(res.status).toBe(200);
+  });
+
+  test('Testing with false values', async () => {
+    const res = await request(app).get(`/query?text=&filter=`);
+    expect(res.status).toBe(400);
+  });
```
Comment (tonyvugithub) [inline test/query.test.js:55]: hey @chawlapalak . This is good. One last thing I want to ask from you. Can you squash all your three commit into one commit with some descriptive comment and force push again? Right now the history of commit has something like fixed fail unit tests. Thank you for your effort.

**Review by tonyvugithub — CHANGES_REQUESTED** [review]

File: test/query.test.js (hunk 8 above)
Comment (chawlapalak) [reply test/query.test.js:55]: Done. Thanks for the feedback and support. Please review the final commit.

**Review by zg3d — DISMISSED** [review]

File: test/query.test.js [hunk 9]
```diff
@@ -19,4 +19,38 @@ describe('Testing query route', () => {
...
+    expect(res.status).toBe(200);
+  });
+
+  test('Testing with valid author filter', async () => {
+    const searchTerm = encodeURIComponent('I Love Telescope');
+    const res = await request(app).get(`/query?text=${searchTerm}&filter=author`);
+    expect(res.status).toBe(200);
+  });
+
+  test('Testing with false values', async () => {
```
Comment (humphd) [inline test/query.test.js:52]: I think `empty` vs. `false` here.

**Review by humphd — CHANGES_REQUESTED** [review]
This is fantastic.  One minor nit to fix, and this is ready to go in.

File: test/query.test.js (hunk 9 above)
Comment (chawlapalak) [reply test/query.test.js]: Hey @humphd , thanks for the feedback. Fixed test descriptions and squashed in one commit. Please review. Thanks.

**Review by humphd — APPROVED** [review]
NOTE to @eunbeek, @zg3d, @tonyvugithub and anyone else working on similar code, when we merge this, others will need to rebase on `master` to pickup the change in package.json vs. doing it over and over again.

**Review by tonyvugithub — APPROVED** [review]

**Pull request merged**